flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/writer/GSRecoverableWriter.java

```diff
@@ -0,0 +1,145 @@
+
+    /** The underlying blob storage. */
+    private final BlobStorage storage;
+
+    /** The GS file system options. */
+    private final GSFileSystemOptions options;
+
+    /**
+     * Construct a GS recoverable writer.
+     *
+     * @param storage The underlying blob storage instance
+     * @param options The GS file system options
+     */
+    public GSRecoverableWriter(BlobStorage storage, GSFileSystemOptions options) {
+        this.storage = Preconditions.checkNotNull(storage);
+        this.options = Preconditions.checkNotNull(options);
+    }
+
+    @Override
+    public boolean requiresCleanupOfRecoverableState() {
+        return true;
+    }
+
+    @Override
+    public boolean supportsResume() {
+        return true;
+    }
+
+    @Override
+    public RecoverableFsDataOutputStream open(Path path) throws IOException {
+        Preconditions.checkNotNull(path);
+
+        BlobId finalBlobId = BlobUtils.parseUri(path.toUri());
+        GSRecoverableWriterState state = new GSRecoverableWriterState(finalBlobId);
+        return new GSRecoverableFsDataOutputStream(storage, options, this, state);
+    }
+
+    @Override
+    public RecoverableFsDataOutputStream recover(ResumeRecoverable resumable) throws IOException {
+        Preconditions.checkNotNull(resumable);
+
+        GSRecoverableWriterState state = (GSRecoverableWriterState) resumable;
+        return new GSRecoverableFsDataOutputStream(storage, options, this, state);
+    }
+
+    @Override
+    public boolean cleanupRecoverableState(ResumeRecoverable resumable) throws IOException {
+        Preconditions.checkNotNull(resumable);
+
+        // determine the partial name for the temporary objects to be deleted
+        GSRecoverableWriterState state = (GSRecoverableWriterState) resumable;
+        String temporaryBucketName = state.getTemporaryBucketName(options);
+        String temporaryObjectPartialName = state.getTemporaryObjectPartialName(options);
+
+        // this will hold the set of blob ids that were actually deleted
+        HashSet<BlobId> deletedBlobIds = new HashSet<>();
+
+        // find all the temp blobs by looking for anything that starts with the temporary
+        // object partial name. doing it this way finds any orphaned temp blobs that might
+        // have come about when resuming
+        List<BlobId> foundTempBlobIds =
+                storage.list(temporaryBucketName, temporaryObjectPartialName);
+        if (!foundTempBlobIds.isEmpty()) {
+
+            // delete all the temp blobs, and populate the set with ones that were actually deleted
+            // normalize in case the blob came back with a generation populated
+            List<Boolean> deleteResults = storage.delete(foundTempBlobIds);
+            for (int i = 0; i < deleteResults.size(); i++) {
+                if (deleteResults.get(i)) {
+                    deletedBlobIds.add(BlobUtils.normalizeBlobId(foundTempBlobIds.get(i)));
+                }
+            }
+        }
```

Comment:
This seems not right.

For each file being written, there could be a series of snapshots (`s1`, `s2`, `s3`, ...) taken before it's committed. Calling `cleanupRecoverableState(s2)` means we no longer need to recover the file writing from `s2`, probably also not likely to recover from `s1` which is taken before `s2`. However, we can still recover from `s3`. Removing all the component blobs of `s2` means some of the component blobs of `s3` are also removed.

Reply:
Oh, I see. I wasn't understanding that ```cleanupRecoverableState``` might be called multiple times during a single recoverable write operation, to clean up some but not all of the temporary blobs -- rather, I was thinking it would just be called once, at the end of a failed recoverable write operation, to clean up everything.

I agree, the scenario you describe could result in needed files being missing. So, to make sure I'm on the same page, is the right thing to do here to just delete those temp blobs directly referenced by the supplied ```ResumeRecoverable```? That would actually simplify some things, e.g. there would be no need to support the ```list``` method on ```BlobStorage``` anymore, the BlobId normalization we discussed elsewhere wouldn't be needed ...

Reply:
Contract wise, `cleanupRecoverableState` is expected to clean-up for a resumable rather than a target file being written. Once a checkpoint is successfully completed, everything happened before that checkpoint is not expected to be replayed. That means there's no need to recover from a resumable generated before that checkpoint, thus the resources for recovering from that resumable can be released.

I think the problem in our case is that, how do we identify temporary blobs that are only referenced by a given resumable. Currently, all the temporary blobs referenced by a previous resumable are also referenced by all subsequent resumables.

Reply:
Reworked [34ad465](https://github.com/apache/flink/pull/15599/commits/34ad465ca419b0afdb399e53deca8b27027f7191) 
according to the [discussion](https://github.com/apache/flink/pull/15599#discussion_r623127365) below.